File: vripper-server/src/main/java/tn/mnlr/vripper/host/ImageBamHost.java
package tn.mnlr.vripper.host;

import lombok.extern.slf4j.Slf4j;
import org.apache.http.client.protocol.HttpClientContext;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.w3c.dom.Document;
import org.w3c.dom.Node;
import tn.mnlr.vripper.exception.HostException;
import tn.mnlr.vripper.exception.XpathException;
import tn.mnlr.vripper.services.HostService;
import tn.mnlr.vripper.services.XpathService;

import java.util.Optional;
import java.util.UUID;

@Service
@Slf4j
public class ImageBamHost extends Host {

  private static final String host = "imagebam.com";
  private static final String IMG_XPATH = "//img[contains(@class,'main-image')]";
  private static final String CONTINUE_XPATH = "//*[contains(text(), 'Continue')]";

  private final HostService hostService;
  private final XpathService xpathService;

  @Autowired
  public ImageBamHost(HostService hostService, XpathService xpathService) {
    this.hostService = hostService;
    this.xpathService = xpathService;
  }

  @Override
  public String getHost() {
    return host;
  }

  @Override
  public String getLookup() {
    return host;
  }

  @Override
  public HostService.NameUrl getNameAndUrl(final String url, final HttpClientContext context)
      throws HostException {

    HostService.Response response = hostService.getResponse(url, context);
    Document doc = response.getDocument();

    try {
      log.debug(String.format("Looking for xpath expression %s in %s", CONTINUE_XPATH, url));
      if (xpathService.getAsNode(doc, CONTINUE_XPATH) != null) {
        // Button detected. No need to actually click it, just make the call again.
        response = hostService.getResponse(url, context);
        doc = response.getDocument();
      }
    } catch (XpathException e) {
      throw new HostException(e);
    }

    Node imgNode;
    try {
      log.debug(String.format("Looking for xpath expression %s in %s", IMG_XPATH, url));
      imgNode = xpathService.getAsNode(doc, IMG_XPATH);
    } catch (XpathException e) {
      throw new HostException(e);
    }

    if (imgNode == null) {
      throw new HostException(String.format("Xpath '%s' cannot be found in '%s'", IMG_XPATH, url));
    }

    try {
      log.debug(String.format("Resolving name and image url for %s", url));
      String imgTitle =
          Optional.ofNullable(imgNode.getAttributes().getNamedItem("alt"))
              .map(e -> e.getTextContent().trim())
              .orElse("");
      String imgUrl =
          Optional.ofNullable(imgNode.getAttributes().getNamedItem("src"))
              .map(e -> e.getTextContent().trim())
              .orElse("");
      String defaultName = UUID.randomUUID().toString();

      int index = imgUrl.lastIndexOf('/');
      if (index != -1 && index < imgUrl.length()) {
        defaultName = imgUrl.substring(imgUrl.lastIndexOf('/') + 1);
      }

      return new HostService.NameUrl(imgTitle.isEmpty() ? defaultName : imgTitle, imgUrl);
    } catch (Exception e) {
      throw new HostException("Unexpected error occurred", e);
    }
  }
}

File: vripper-core/src/main/kotlin/me/vripper/host/ImageBamHost.kt
package me.vripper.host

import me.vripper.download.ImageDownloadContext
import me.vripper.exception.HostException
import me.vripper.exception.XpathException
import me.vripper.services.*
import me.vripper.utilities.HtmlUtils
import me.vripper.utilities.XpathUtils
import org.apache.hc.client5.http.impl.cookie.BasicClientCookie
import org.w3c.dom.Document
import org.w3c.dom.Node
import java.sql.Date
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.*

class ImageBamHost(
    httpService: HTTPService,
    dataTransaction: DataTransaction,
    downloadSpeedService: DownloadSpeedService,
) : Host("imagebam.com", 2, httpService, dataTransaction, downloadSpeedService) {
    private val log by me.vripper.delegate.LoggerDelegate()

    @Throws(HostException::class)
    override fun resolve(
        url: String,
        document: Document,
        context: ImageDownloadContext
    ): Pair<String, String> {
        val doc = try {
            log.debug(String.format("Looking for xpath expression %s in %s", CONTINUE_XPATH, url))
            if (XpathUtils.getAsNode(document, CONTINUE_XPATH) != null) {
                val clientCookie = BasicClientCookie("nsfw_inter", "1")
                clientCookie.domain = "www.imagebam.com"
                clientCookie.path = "/"
                clientCookie.expiryDate =
                    Date.from(
                        LocalDateTime.now().plusDays(3).atZone(ZoneId.systemDefault()).toInstant()
                    )
                context.httpContext.cookieStore.addCookie(clientCookie)
                fetch(url, context) {
                    HtmlUtils.clean(it.entity.content)
                }
            } else {
                document
            }
        } catch (e: XpathException) {
            throw HostException(e)
        }
        val imgNode: Node = try {
            log.debug(String.format("Looking for xpath expression %s in %s", IMG_XPATH, url))
            XpathUtils.getAsNode(doc, IMG_XPATH)
        } catch (e: XpathException) {
            throw HostException(e)
        } ?: throw HostException(
            String.format(
                "Xpath '%s' cannot be found in '%s'",
                IMG_XPATH,
                url
            )
        )
        return try {
            log.debug(String.format("Resolving name and image url for %s", url))
            val imgTitle = Optional.ofNullable(imgNode.attributes.getNamedItem("alt"))
                .map { e: Node -> e.textContent.trim { it <= ' ' } }
                .orElse("")
            val imgUrl = Optional.ofNullable(imgNode.attributes.getNamedItem("src"))
                .map { e: Node -> e.textContent.trim { it <= ' ' } }
                .orElse("")
            var defaultName: String = UUID.randomUUID().toString()
            val index = imgUrl.lastIndexOf('/')
            if (index != -1 && index < imgUrl.length) {
                defaultName = imgUrl.substring(imgUrl.lastIndexOf('/') + 1)
            }
            Pair((imgTitle.ifEmpty { defaultName })!!, imgUrl)
        } catch (e: Exception) {
            throw HostException("Unexpected error occurred", e)
        }
    }

    companion object {
        private const val IMG_XPATH = "//img[contains(@class,'main-image')]"
        private const val CONTINUE_XPATH = "//*[contains(text(), 'Continue')]"
    }
}